File: BaekJoon/15664_N과M(10)_BackTracking.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

// N과M(10) 15664 BackTracking
// O( M * N^M )


int main(void)
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);

    int n, m;
    cin >> n >> m;

    vector<int> v(n);
    for (int& e : v)
    {
        cin >> e;
    }

    // 사전순 처리와 중복 방지를 위한 오름차순 정렬
    sort(v.begin(), v.end());

    vector<int> seq;

    auto DFS = [&](auto&& self, int start) -> void {
        // m 개를 선택했다면 출력
        if (seq.size() == m)
        {
            for (const int& e : seq)
            {
                cout << e << " ";
            }
            cout << "\n";
            return;
        }

        for (int i = start; i < n; ++i)
        {
            // 동일 깊이에서 중복 값 건너뜀
            if (i > start && v[i] == v[i - 1]) continue;

            seq.push_back(v[i]);
            self(self, i + 1);
            seq.pop_back(); // 백트래킹
        }
        };

    DFS(DFS, 0);


    return 0;
}

File: BaekJoon/Java/15664_N과M(10)_BackTracking.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

// N과M(10) 15664 BackTracking
// O( M * N^M )


public class Main {
    static int n, m;
    static int[] nums, seq;
    static StringBuilder sb = new StringBuilder();

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());

        nums = new int[n];
        seq = new int[m];
        st = new StringTokenizer(br.readLine());

        for (int i = 0; i < n; ++i) {
            nums[i] = Integer.parseInt(st.nextToken());
        }

        Arrays.sort(nums);

        DFS(0, 0);

        System.out.println(sb);
    }

    private static void DFS(int depth, int start) {
        if (depth == m) {
            for (int i = 0; i < m; ++i) {
                sb.append(seq[i]).append(" ");
            }
            sb.append("\n");
            return;
        }

        int prev = -1;
        for (int i = start; i < n; ++i) {
            if (prev == nums[i]) continue;

            seq[depth] = nums[i];
            DFS(depth + 1, i + 1);
            prev = nums[i];
        }
    }
}
